### PR TITLE
*: fix bugs that creating index on clustered index table or table with the virtual generated column

### LINKED ISSUE #18061: cannot create index on virtual generated column.
## Bug Report

 Cannot create index on virtual generated column gened by primary key after #18045 merge.

### 1. Minimal reproduce step (Required)
```
drop database if exists `db`;
create database `db`;
use `db`;
create table t1 (id int, PRIMARY KEY (id));
insert into t1 values(1);
alter table t1 add column gen_id int as (id+1);
alter table t1 add index idx(`gen_id`);
```

### 2. What did you expect to see? (Required)
no error
### 3. What did you see instead (Required)
```
ERROR 8202 (HY000): Cannot decode index value, because Miss column
```
### 4. Affected version (Required)
Server version: 5.7.25-TiDB-v4.0.0-beta.2-636-gc19018e06

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve?

Issue Number: close #18061 

Problem Summary:
I wrongly remove the handle's decode logic in #18045.
Anyway, the previously handle's decode logic didn't consider common handle, it's not completely correct. 

### What is changed and how it works?

1. Add the handle's decode logic back in `DecodeAndEvalRowWithMap()` and implement the common handle part.
2. Fix the wrong decode logic for common handle in `decodeOldRowValToChunk()`, point_get.go
3. Remove the handle's decode logic after the `DecodeAndEvalRowWithMap()` , do it in `DecodeAndEvalRowWithMap()`. 
4. Wrap two functions `TryGetCommonPkColumnIds` and `TryGetCommonPkColumns`.

### Check List 

Tests 

- Unit test

Side effects

### Release note 

- No release note

### CODE REVIEW

File: ddl/db_test.go [hunk 1]
```diff
@@ -3752,6 +3770,17 @@ func (s *testDBSuite5) TestAddIndexForGeneratedColumn(c *C) {
 	s.tk.MustQuery("select id1 from gcai_table use index(idx1)").Check(testkit.Rows("6"))
 	s.tk.MustExec("admin check table gcai_table")
 }
+func (s *testSerialDBSuite) TestAddIndexForGeneratedColumn(c *C) {
+	s.tk = testkit.NewTestKit(c, s.store)
+	save := config.GetGlobalConfig().AlterPrimaryKey
+	config.GetGlobalConfig().AlterPrimaryKey = false
```
Comment (tangenta) [inline ddl/db_test.go:3776]: https://github.com/pingcap/tidb/pull/17964 This PR provides a way to update the global config without data race. Perhaps we should comply with it.

```
config.UpdateGlobal(func(conf *config.Config) {
		conf.AlterPrimaryKey = true
	})
defer config.RestoreFunc()()
```

File: ddl/db_test.go [hunk 2]
```diff
@@ -139,27 +139,44 @@ type testDBSuite6 struct{ *testDBSuite }
...
+	s.tk.MustExec("create table t (a int, b int, c int, primary key(a, b));")
+	s.tk.MustExec("insert into t values (1, 2, 3);")
+	s.tk.MustExec("create index idx on t (a, b);")
+}
+
+func (s *testSerialDBSuite) TestAddIndexWithPK(c *C) {
+	s.tk = testkit.NewTestKit(c, s.store)
+	s.tk.MustExec("use " + s.schemaName)
+	save := config.GetGlobalConfig().AlterPrimaryKey
+	config.GetGlobalConfig().AlterPrimaryKey = false
```
Comment (tangenta) [inline ddl/db_test.go:173]: Ditto

File: ddl/db_test.go (hunk 1 above)
Comment (wjhuang2016) [reply ddl/db_test.go:3776]: Sure

File: util/rowDecoder/decoder.go [hunk 3]
```diff
@@ -88,17 +120,21 @@ func (rd *RowDecoder) DecodeAndEvalRowWithMap(ctx sessionctx.Context, handle kv.
 	if err != nil {
 		return nil, err
 	}
-	if !rd.haveGenColumn {
-		return row, nil
-	}
-
+	pkCols := tables.TryGetCommonPkColumnIds(tbl)
```
Comment (coocood) [inline util/rowDecoder/decoder.go:123]: The RowDecoder already has a table field, we show create pkCols in NewRowDecoder.

File: util/rowDecoder/decoder.go [hunk 4]
```diff
@@ -74,11 +72,47 @@ func NewRowDecoder(tbl table.Table, decodeColMap map[int64]Column) *RowDecoder {
...
+				return true, nil
+			}
+		}
+	}
+	return false, nil
+}
+
 // DecodeAndEvalRowWithMap decodes a byte slice into datums and evaluates the generated column value.
-func (rd *RowDecoder) DecodeAndEvalRowWithMap(ctx sessionctx.Context, handle kv.Handle, b []byte, decodeLoc, sysLoc *time.Location, row map[int64]types.Datum) (map[int64]types.Datum, error) {
+func (rd *RowDecoder) DecodeAndEvalRowWithMap(ctx sessionctx.Context, handle kv.Handle, b []byte, decodeLoc, sysLoc *time.Location, row map[int64]types.Datum, tbl *model.TableInfo) (map[int64]types.Datum, error) {
```
Comment (coocood) [inline util/rowDecoder/decoder.go:115]: The `tbl` arg can be removed.

**Comment by coocood** [comment]
LGTM

**Pull request merged**